contracts/external/cw-verifier-middleware/src/verify.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{
+    error::ContractError,
+    msg::WrappedMessage,
+    state::{CONTRACT_ADDRESS, NONCES},
+};
+use bech32::{ToBase32, Variant};
+use cosmwasm_schema::{schemars::_serde_json::json, serde::de::DeserializeOwned};
+use cosmwasm_std::{
+    from_slice, to_binary, Addr, Api, DepsMut, Env, MessageInfo, StdError, Storage, Uint128,
+};
+
+use ripemd::Ripemd160;
+use sha2::{Digest, Sha256};
+
+const UNCOMPRESSED_HEX_PK_LEN: usize = 130;
+const COMPRESSED_HEX_PK_LEN: usize = 66;
+
+pub fn verify<T>(
+    api: &dyn Api,
+    storage: &mut dyn Storage,
+    env: &Env,
+    info: MessageInfo,
+    wrapped_msg: WrappedMessage,
+) -> Result<(T, MessageInfo), ContractError>
+where
+    T: DeserializeOwned,
+{
+    let payload = wrapped_msg.payload;
+
+    let signer_addr = pk_to_addr(
+        api,
+        wrapped_msg.public_key.to_hex(), // to_hex ensures that the public key has the expected number of bytes
+        &payload.bech32_prefix,
+    )?;
+
+    let payload_ser = serde_json::to_string(&payload)?;
```

Comment:
it's a bit funny that we have to use the deserialized bech32_prefix, and also reserialize to get its signed representation, but i can think of no other way. the quirks of the universe.